### ISSUE TITLE
[BUG] Rancher 2.7.1 - vSphere provisioning downstream cluster node stuck at "waiting for SSH"

### ISSUE THREAD

**Issue by davidhrbac** [rancher/rancher#40917]
**Rancher Server Setup**
- Rancher version: 2.7.1
- Installation option (Docker install/Helm Chart): Helm
   - If Helm Chart, Kubernetes Cluster and version (RKE1, RKE2, k3s, EKS, etc): RKE 1.4.0, 1.26.4
- Proxy/Cert Details: None

vSphere: 7.0.3 Build: 20990077

**User Information**
- What is the role of the user logged in? full admin

**Describe the bug**
Brand new 2.7.1 Rancher local cluster running atop of v1.24.6 K8s. When trying to create a new downstream cluster (v1.24.10-rancher4-1, 3 nodes) with vSphere template, VMs are created, docker user is created by cloud init. Process is stuck on waiting for SSH connection to test. We have tested manually that SSH connection is working within the container and user and SSH key is OK. It seems that provisioner should be using `--native-ssh` to work again?

**To Reproduce**
We have repeated the situation multiple times.

**Result**

```console
Waiting for machine to be running, this may take a few minutes...
(pjarka555) Calling .GetState
(pjarka555) DBG | 2023/03/17 10:38:55 1-0217.req.headers
(pjarka555) DBG | 2023/03/17 10:38:55 POST /sdk HTTP/1.1
(pjarka555) DBG | Host: 10.14.48.10:443
(pjarka555) DBG | Content-Type: text/xml; charset="utf-8"
(pjarka555) DBG | Soapaction: urn:vim25/7.0
(pjarka555) DBG |
(pjarka555) DBG | 2023/03/17 10:38:55 1-0217.req.xml
(pjarka555) DBG | 2023/03/17 10:38:55 <?xml version="1.0" encoding="UTF-8"?>(pjarka555) DBG | 2023/03/17 10:38:55 <Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/"><Body><RetrieveProperties xmlns="urn:vim25"><_this type="PropertyCollector">propertyCollector</_this><specSet><propSet><type>VirtualMachine</type><all>true</all></propSet><objectSet><obj type="VirtualMachine">vm-64619</obj><skip>false</skip></objectSet></specSet></RetrieveProperties></Body></Envelope>(pjarka555) DBG | 2023/03/17 10:38:55 2023-03-17T10-38-55.854335569 - 0217:
(pjarka555) DBG | 2023/03/17 10:38:55      3ms (*methods.RetrievePropertiesBody)
(pjarka555) DBG | 2023/03/17 10:38:55
(pjarka555) DBG | 2023/03/17 10:38:55 1-0217.res.headers
(pjarka555) DBG | 2023/03/17 10:38:55 HTTP/1.1 200 OK
(pjarka555) DBG | Transfer-Encoding: chunked
(pjarka555) DBG | Cache-Control: no-cache
(pjarka555) DBG | Content-Type: text/xml; charset=utf-8
(pjarka555) DBG | Date: Fri, 17 Mar 2023 10:38:55 GMT
(pjarka555) DBG | Server: envoy
(pjarka555) DBG | X-Applied-Soapaction: urn:vim25/7.0.3.0
(pjarka555) DBG | X-Envoy-Upstream-Service-Time: 1
(pjarka555) DBG |
(pjarka555) DBG | 2023/03/17 10:38:55 1-0217.res.xml
(pjarka555) DBG | 2023/03/17 10:38:55 <?xml version="1.0" encoding="UTF-8"?>(pjarka555) DBG | <soapenv:Envelope xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/"
(pjarka555) DBG |  xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
(pjarka555) DBG |  xmlns:xsd="http://www.w3.org/2001/XMLSchema"
Waiting for SSH to be available...

...

Getting to WaitForSSH function...
(pjarka555) Calling .GetSSHHostname
(pjarka555) DBG | 2023/03/17 10:38:55 /hostMemoryUsage><guestHeartbeatStatus>green</guestHeartbeatStatus><distributedCpuEntitlement>4200</distributedCpuEntitlement><distributedMemoryEntitlement>16733</distributedMemoryEntitlement><staticCpuEntitlement>1825</staticCpuEntitlement><staticMemoryEntitlement>16733</staticMemoryEntitlement><grantedMemory>16384</grantedMemory><privateMemory>16384</privateMemory><sharedMemory>0</sharedMemory><swappedMemory>0</swappedMemory><balloonedMemory>0</balloonedMemory><consumedOverheadMemory>0</consumedOverheadMemory><ftLogBandwidth>-1</ftLogBandwidth><ftSecondaryLatency>-1</ftSecondaryLatency><ftLatencyStatus>gray</ftLatencyStatus><compressedMemory>0</compressedMemory><uptimeSeconds>46</uptimeSeconds><ssdSwappedMemory>0</ssdSwappedMemory><activeMemory>16384</activeMemory></quickStats><overallStatus>green</overallStatus></val></propSet><propSet><name>tag</name><val xsi:type="ArrayOfTag"></val></propSet><propSet><name>triggeredAlarmState</name><val xsi:type="ArrayOfAlarmState"></val></propSet><propSet><name>value</name><val xsi:type="ArrayOfCustomFieldValue"></val></propSet></returnval></RetrievePropertiesResponse>(pjarka555) DBG | </soapenv:Body>(pjarka555) DBG | </soapenv:Envelope>(pjarka555) Calling .GetSSHPort
(pjarka555) Calling .GetSSHKeyPath
(pjarka555) Calling .GetSSHKeyPath
(pjarka555) Calling .GetSSHUsername
Using SSH client type: external
proxy_url: ; ncBinaryPath: /usr/bin/nc
Using SSH private key: /root/.docker/machine/machines/pjarka555/id_rsa (-rw-------)
&{[-F /dev/null -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none -o LogLevel=quiet -o PasswordAuthentication=no -o ServerAliveInterval=60 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null docker@10.14.17.164 -o IdentitiesOnly=yes -i /root/.docker/machine/machines/pjarka555/id_rsa -p 22] /usr/bin/ssh <nil>}
About to run SSH command:
exit 0
```

Provisioning node is overloaded with SSH processes trying to communicate:

```console
[root@hrb-vm-1 ~]# ps aux | grep "ssh -F"
root     2978152 88.4  0.0  37780  5772 ?        R    06:33  14:38 /usr/bin/ssh -F /dev/null -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none -o LogLevel=quiet -o PasswordAuthentication=no -o ServerAliveInterval=60 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null docker@10.14.17.147 -o IdentitiesOnly=yes -i /management-state/node/nodes/pjarka1/machines/pjarka1/id_rsa -p 22 exit 0
root     2978206 88.3  0.0  37780  5708 ?        R    06:33  14:33 /usr/bin/ssh -F /dev/null -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none -o LogLevel=quiet -o PasswordAuthentication=no -o ServerAliveInterval=60 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null docker@10.14.17.148 -o IdentitiesOnly=yes -i /management-state/node/nodes/pjarka2/machines/pjarka2/id_rsa -p 22 exit 0
root     2978513 88.2  0.0  37780  5804 ?        R    06:34  14:26 /usr/bin/ssh -F /dev/null -o ConnectionAttempts=3 -o ConnectTimeout=10 -o ControlMaster=no -o ControlPath=none -o LogLevel=quiet -o PasswordAuthentication=no -o ServerAliveInterval=60 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null docker@10.14.17.149 -o IdentitiesOnly=yes -i /management-state/node/nodes/pjarka3/machines/pjarka3/id_rsa -p 22 exit 0
```

We have identified with `tcpdump` there is no communication on SSH towards new nodes.

It does not work:
```console
sh-4.4# rancher-machine -D create -d vmwarevsphere --engine-install-url https://releases.rancher.com/install-docker/20.10.sh --vmwarevsphere-clone-from /TIECZOSRLDC/vm/Cuberneties/template-rockylinux9-0.0.2 --vmwarevsphere-memory-size 16384 --vmwarevsphere-ssh-user-group staff --vmwarevsphere-vcenter 10.14.48.10 --vmwarevsphere-disk-size 100000 --vmwarevsphere-network /TIECZOSRLDC/network/912_PROD_TIECZOSR_10.14.17.128m26 --vmwarevsphere-password XXXXX --vmwarevsphere-ssh-password tcuser --vmwarevsphere-creation-type template --vmwarevsphere-datacenter /TIECZOSRLDC --vmwarevsphere-datastore /TIECZOSRLDC/datastore/TIECZOSRLDS_PROD1 --vmwarevsphere-pool /TIECZOSRLDC/host/TIECZOSRLCLU_PROD1/Resources --vmwarevsphere-username kubernetes@vsphere.local --vmwarevsphere-ssh-port 22 --vmwarevsphere-ssh-user docker --vmwarevsphere-cfgparam disk.enableUUID=TRUE --vmwarevsphere-cpu-count 2 --vmwarevsphere-folder /TIECZOSRLDC/vm/Cuberneties --vmwarevsphere-vcenter-port 443 --vmwarevsphere-cloud-config /RDEVSVNQER/cloud-config --vmwarevsphere-os linux pjarka555
```

Adding `--native-ssh` seems to help process to go:

```console
sh-4.4# rancher-machine --native-ssh -D create -d vmwarevsphere --engine-install-url https://releases.rancher.com/install-docker/20.10.sh --vmwarevsphere-clone-from /TIECZOSRLDC/vm/Cuberneties/template-rockylinux9-0.0.2 --vmwarevsphere-memory-size 16384 --vmwarevsphere-ssh-user-group staff --vmwarevsphere-vcenter 10.14.48.10 --vmwarevsphere-disk-size 100000 --vmwarevsphere-network /TIECZOSRLDC/network/912_PROD_TIECZOSR_10.14.17.128m26 --vmwarevsphere-password XXXXX --vmwarevsphere-ssh-password tcuser --vmwarevsphere-creation-type template --vmwarevsphere-datacenter /TIECZOSRLDC --vmwarevsphere-datastore /TIECZOSRLDC/datastore/TIECZOSRLDS_PROD1 --vmwarevsphere-pool /TIECZOSRLDC/host/TIECZOSRLCLU_PROD1/Resources --vmwarevsphere-username kubernetes@vsphere.local --vmwarevsphere-ssh-port 22 --vmwarevsphere-ssh-user docker --vmwarevsphere-cfgparam disk.enableUUID=TRUE --vmwarevsphere-cpu-count 2 --vmwarevsphere-folder /TIECZOSRLDC/vm/Cuberneties --vmwarevsphere-vcenter-port 443 --vmwarevsphere-cloud-config /RDEVSVNQER/cloud-config --vmwarevsphere-os linux pjarka555
```

And the log seems OK now:

```console
Waiting for machine to be running, this may take a few minutes...
(pjarka444) Calling .GetState
(pjarka444) DBG | 2023/03/17 11:02:37 1-0220.req.headers
(pjarka444) DBG | 2023/03/17 11:02:37 POST /sdk HTTP/1.1
(pjarka444) DBG | Host: 10.14.48.10:443
(pjarka444) DBG | Content-Type: text/xml; charset="utf-8"
(pjarka444) DBG | Soapaction: urn:vim25/7.0
(pjarka444) DBG |
(pjarka444) DBG | 2023/03/17 11:02:37 1-0220.req.xml
(pjarka444) DBG | 2023/03/17 11:02:37 <?xml version="1.0" encoding="UTF-8"?>(pjarka444) DBG | 2023/03/17 11:02:37 <Envelope xmlns="http://schemas.xmlsoap.org/soap/envelope/"><Body><RetrieveProperties xmlns="urn:vim25"><_this type="PropertyCollector">propertyCollector</_this><specSet><propSet><type>VirtualMachine</type><all>true</all></propSet><objectSet><obj type="VirtualMachine">vm-64620</obj><skip>false</skip></objectSet></specSet></RetrieveProperties></Body></Envelope>(pjarka444) DBG | 2023/03/17 11:02:37 2023-03-17T11-02-37.508216235 - 0220:
(pjarka444) DBG | 2023/03/17 11:02:37      2ms (*methods.RetrievePropertiesBody)
(pjarka444) DBG | 2023/03/17 11:02:37
(pjarka444) DBG | 2023/03/17 11:02:37 1-0220.res.headers
(pjarka444) DBG | 2023/03/17 11:02:37 HTTP/1.1 200 OK
(pjarka444) DBG | Transfer-Encoding: chunked
(pjarka444) DBG | Cache-Control: no-cache
(pjarka444) DBG | Content-Type: text/xml; charset=utf-8
(pjarka444) DBG | Date: Fri, 17 Mar 2023 11:02:37 GMT
(pjarka444) DBG | Server: envoy
(pjarka444) DBG | X-Applied-Soapaction: urn:vim25/7.0.3.0
(pjarka444) DBG | X-Envoy-Upstream-Service-Time: 1
(pjarka444) DBG |
(pjarka444) DBG | 2023/03/17 11:02:37 1-0220.res.xml
(pjarka444) DBG | 2023/03/17 11:02:37 <?xml version="1.0" encoding="UTF-8"?>(pjarka444) DBG | <soapenv:Envelope xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/"
(pjarka444) DBG |  xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/"
(pjarka444) DBG |  xmlns:xsd="http://www.w3.org/2001/XMLSchema"

... 


About to run SSH command:
if grep -xq 127.0.1.1.* /etc/hosts; then sudo sed -i 's/^127.0.1.1.*/127.0.1.1 pjarka444/g' /etc/hosts; else echo '127.0.1.1 pjarka444' | sudo tee -a /etc/hosts; fi
SSH cmd err, output: <nil>: 127.0.1.1 pjarka444
installing base package: name=curl
(pjarka444) Calling .GetSSHHostname
(pjarka444) Calling .GetSSHPort
(pjarka444) Calling .GetSSHKeyPath
(pjarka444) Calling .GetSSHKeyPath
(pjarka444) Calling .GetSSHUsername
Using SSH client type: native
&{{{<nil> 0 [] [] []} docker [0x7ca5a0] 0x7ca560 <nil>  [] 0s} 10.14.17.165 22 <nil> <nil>}
About to run SSH command:
sudo -E yum install -y curl
SSH cmd err, output: <nil>: Rocky Linux 9 - BaseOS                                                                                                   [===                                                                                                                 ] --- Rocky Linux 9 - BaseOS                                                                                                                                                                                                                          8.8 kB/s | 4.1 kB     00:00
Rocky Linux 9 - BaseOS                                                                                                                                                                                                                          1.7 MB/s | 1.8 MB     00:01
Rocky Linux 9 - AppStream                                                                                                                                                                                                                        11 kB/s | 4.5 kB     00:00
Rocky Linux 9 - Extras                                                                                                                                                                                                                          6.7 kB/s | 2.9 kB     00:00
Package curl-7.76.1-19.el9_1.1.x86_64 is already installed.
Dependencies resolved.
Nothing to do.
Complete!
Installing Docker from: https://releases.rancher.com/install-docker/20.10.sh
```

**Expected Result**
New downstream cluster should be created.

**Comment by mpepping** [comment]
Same behavior here after upgrading to v2.6.11.

**Comment by davidhrbac** [comment]
I have been working on that my last 48h. I thought the issue is there because of me using Rocky 9, so I stepped down to Rocky 8. It was the same. So I went different routes and investigated.

The root cause was missing Rancher host resolving on downstream nodes. Here we go again, **Rancher host resolving by having the record in `/etc/hosts` is not enough. DNS server must be providing the Rancher host record for `rancher-agent` is _not using_ `/etc/hosts` on the node!**

And since I have already been playing with vSphere integration, I have tested that Rancher 2.7.1:

- provisioning of RKE cluster on vSphere is working with Rocky 8,
- provisioning of RKE2 cluster on vSphere is working with Rocky 8,
- provisioning of RKE cluster on vSphere is working with Rocky 9,
- provisioning of RKE2 cluster on vSphere is working with Rocky 9,
- Terraform Rancher2 provider creates RKE cluster with Rocky 8,
- Terraform Rancher2 provider creates RKE cluster with Rocky 9,
- I did not test Terraform Rancher2 provider to create RKE2.

![image](https://user-images.githubusercontent.com/728971/226198530-025a014d-04ad-49e8-8fa1-3c80376a9373.png)

I'm sorry to have created the issue. I think we can close the ticket.
